File: apps/somma-expenses/backend/internal/jobs/budget_alerts.go
package jobs

import (
	"log"
	"time"
)

func (s *Scheduler) checkBudgetAlerts() {
	now := time.Now().UTC()
	today := now.Format("2006-01-02")

	rows, err := s.db.Query(s.ctx, `
		SELECT b.id, b.user_id, b.name,
		       COALESCE(b.allocated_amount_cents, 0) AS total_cents,
		       COALESCE(s.spent_cents, 0) AS spent_cents
		FROM budgets b
		LEFT JOIN (
			SELECT budget_id,
			       SUM(CASE WHEN type = 'EXPENSE' THEN amount_cents ELSE 0 END) AS spent_cents
			FROM transactions
			WHERE is_active = true
			  AND status = 'COMPLETED'
			  AND budget_id IS NOT NULL
			GROUP BY budget_id
		) s ON s.budget_id = b.id
		WHERE b.is_active = true
		  AND COALESCE(b.allocated_amount_cents, 0) > 0
		  AND COALESCE(s.spent_cents, 0) >= COALESCE(b.allocated_amount_cents, 0) * 0.80
		ORDER BY b.name ASC
	`)
	if err != nil {
		log.Printf("jobs: budget alerts query error: %v", err)
		return
	}
	defer rows.Close()

	type budgetRow struct {
		id         string
		userID     string
		name       string
		totalCents int64
		spentCents int64
	}
	var budgets []budgetRow
	for rows.Next() {
		var b budgetRow
		if err := rows.Scan(&b.id, &b.userID, &b.name, &b.totalCents, &b.spentCents); err != nil {
			log.Printf("jobs: budget alert scan error: %v", err)
			continue
		}
		budgets = append(budgets, b)
	}
	rows.Close()

	for _, b := range budgets {
		percent := float64(b.spentCents) / float64(b.totalCents) * 100

		thresholds := []int{80, 100}
		for _, threshold := range thresholds {
			if percent < float64(threshold) {
				continue
			}
			var exists bool
			err := s.db.QueryRow(s.ctx, `
				SELECT EXISTS(
					SELECT 1 FROM budget_alerts
					WHERE budget_id = $1
					  AND threshold_pct = $2
					  AND created_at::date = $3::date
				)
			`, b.id, threshold, today).Scan(&exists)
			if err != nil {
				log.Printf("jobs: budget alert debounce check error: %v", err)
				continue
			}
			if exists {
				continue
			}
			_, err = s.db.Exec(s.ctx, `
				INSERT INTO budget_alerts
					(user_id, budget_id, budget_name, threshold_pct, percent_used, allocated_cents, spent_cents)
				VALUES ($1, $2, $3, $4, $5, $6, $7)
			`, b.userID, b.id, b.name, threshold, percent, b.totalCents, b.spentCents)
			if err != nil {
				log.Printf("jobs: budget alert insert error budget=%s threshold=%d: %v", b.id, threshold, err)
			}
		}
	}
}

File: apps/somma-expenses/backend/src/jobs/budget_alerts.rs
use chrono::Utc;
use sqlx::{PgPool, Row};
use tracing::error;
use uuid::Uuid;

pub async fn check_budget_alerts(db: &PgPool) {
    let now = Utc::now();
    let today = now.date_naive();

    struct BudgetRow {
        id: Uuid,
        user_id: Uuid,
        name: String,
        total_cents: i64,
        spent_cents: i64,
    }

    let rows = match sqlx::query(
        r#"
        SELECT b.id, b.user_id, b.name,
               COALESCE(b.allocated_amount_cents, 0)::bigint AS total_cents,
               COALESCE(s.spent_cents, 0)::bigint AS spent_cents
        FROM budgets b
        LEFT JOIN (
            SELECT budget_id,
                   SUM(CASE WHEN type = 'EXPENSE' THEN amount_cents ELSE 0 END)::bigint AS spent_cents
            FROM transactions
            WHERE is_active = true
              AND status = 'COMPLETED'
              AND budget_id IS NOT NULL
            GROUP BY budget_id
        ) s ON s.budget_id = b.id
        WHERE b.is_active = true
          AND COALESCE(b.allocated_amount_cents, 0) > 0
          AND COALESCE(s.spent_cents, 0) >= COALESCE(b.allocated_amount_cents, 0) * 0.80
        ORDER BY b.name ASC
        "#
    )
    .fetch_all(db)
    .await
    {
        Ok(r) => r,
        Err(e) => {
            error!("jobs: budget alerts query error: {}", e);
            return;
        }
    };

    let budgets: Vec<BudgetRow> = rows
        .into_iter()
        .map(|r| BudgetRow {
            id: r.get("id"),
            user_id: r.get("user_id"),
            name: r.get("name"),
            total_cents: r.get("total_cents"),
            spent_cents: r.get("spent_cents"),
        })
        .collect();

    for b in budgets {
        let percent = (b.spent_cents as f64 / b.total_cents as f64) * 100.0;
        let thresholds = [80, 100];

        for threshold in thresholds {
            if percent < threshold as f64 {
                continue;
            }

            let exists_row = sqlx::query(
                r#"
                SELECT EXISTS(
                    SELECT 1 FROM budget_alerts
                    WHERE budget_id = $1
                      AND threshold_pct = $2
                      AND created_at::date = $3
                ) as exists
                "#,
            )
            .bind(b.id)
            .bind(threshold)
            .bind(today)
            .fetch_one(db)
            .await;

            let exists = match exists_row {
                Ok(r) => r.try_get("exists").unwrap_or(false),
                Err(_) => false,
            };

            if exists {
                continue;
            }

            let _ = sqlx::query(
                r#"
                INSERT INTO budget_alerts
                    (user_id, budget_id, budget_name, threshold_pct, percent_used, allocated_cents, spent_cents)
                VALUES ($1, $2, $3, $4, $5, $6, $7)
                "#,
            )
            .bind(b.user_id)
            .bind(b.id)
            .bind(b.name.clone())
            .bind(threshold)
            .bind(percent)
            .bind(b.total_cents)
            .bind(b.spent_cents)
            .execute(db)
            .await;
        }
    }
}
